File: sol_1_6_4.c
#include <stdio.h>
#include <string.h>
#include <unistd.h> /* optarg */
#include "defs.h"


int _debug = DEBUG_OFF;


struct lcd
{
    int  _base;
    char _digits[10];
};
typedef struct lcd lcd_t;


void print_zero(int base, int row, int column)
{
    if (row == 0 || row == 2*base+2)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else if (row == base+1)
        printf(" ");
    else
    {
        if (column == 0 || column == base+1)
            printf("|");
        else
            printf(" ");
    }
}


void print_one(int base, int row, int column)
{
    if (row == 0 || row == base+1 || row == 2*base+2)
        printf(" ");
    else
    {
        if (column == base+1)
            printf("|");
        else
            printf(" ");
    }
}


void print_two(int base, int row, int column)
{
    if (row == 0 || row == base+1 || row == 2*base+2)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else
    {
        if (row <= base)
        {
            if (column == base+1)
                printf("|");
            else
                printf(" ");
        }
        else
        {
            if (column == 0)
                printf("|");
            else
                printf(" ");
        }
    }
}


void print_three(int base, int row, int column)
{
    if (row == 0 || row == base+1 || row == 2*base+2)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else
    {
        if (column == base+1)
            printf("|");
        else
            printf(" ");
    }
}


void print_four(int base, int row, int column)
{
    if (row == 0 || row == 2*base+2)
        printf(" ");
    else if (row == base+1)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else
    {
        if (row <= base)
        {
            if (column == 0 || column == base+1)
                printf("|");
            else
                printf(" ");
        }
        else
        {
            if (column == base+1)
                printf("|");
            else
                printf(" ");
        }
    }
}


void print_five(int base, int row, int column)
{
    if (row == 0 || row == base+1 || row == 2*base+2)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else
    {
        if (row <= base)
        {
            if (column == 0)
                printf("|");
            else
                printf(" ");
        }
        else
        {
            if (column == base+1)
                printf("|");
            else
                printf(" ");
        }
    }
}


void print_six(int base, int row, int column)
{
    if (row == 0 || row == base+1 || row == 2*base+2)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else
    {
        if (row <= base)
        {
            if (column == 0)
                printf("|");
            else
                printf(" ");
        }
        else
        {
            if (column == 0 || column == base+1)
                printf("|");
            else
                printf(" ");
        }
    }
}


void print_seven(int base, int row, int column)
{
    if (row == 0)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else if (row == base+1 || row == 2*base+2)
        printf(" ");
    else
    {
        if (column == base+1)
            printf("|");
        else
            printf(" ");
    }
}


void print_eight(int base, int row, int column)
{
    if (row == 0 || row == base+1 || row == 2*base+2)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else
    {
        if (column == 0 || column == base+1)
            printf("|");
        else
            printf(" ");
    }
}


void print_nine(int base, int row, int column)
{
    if (row == 0 || row == base+1 || row == 2*base+2)
    {
        if (column == 0 || column == base+1)
            printf(" ");
        else
            printf("-");
    }
    else
    {
        if (row <= base)
        {
            if (column == 0 || column == base+1)
                printf("|");
            else
                printf(" ");
        }
        else
        {
            if (column == base+1)
                printf("|");
            else
                printf(" ");
        }
    }
}


void print(int base, char* digits)
{
    /* define and initialize an array of pointer to print functions */
    static void (*prints[10])(int, int, int);
    static int init = 0;

    if (init == 0)
    {
        prints[0] = &print_zero;
        prints[1] = &print_one;
        prints[2] = &print_two;
        prints[3] = &print_three;
        prints[4] = &print_four;
        prints[5] = &print_five;
        prints[6] = &print_six;
        prints[7] = &print_seven;
        prints[8] = &print_eight;
        prints[9] = &print_nine;

        init = 1;
    }


    if (_debug >= DEBUG_COPIOUS)
        printf("base: %d, digits: %s, digits length: %d\n",
               base, digits, strlen(digits));

    int row;
    for (row = 0; row < 2*base+3; row++)
    {
        int i;
        for (i = 0; i < strlen(digits); i++)
        {
            if (i != 0)
                printf(" ");
            int column;
            for (column = 0; column < base+2; column++)
            {
                int digit = digits[i] - '0';
                (*prints[digit])(base, row, column);
            }
        }
        printf("\n");
    }
}

 
int input(char* input_file, lcd_t* lcds, int* count)
{
    int success;
    FILE* file = fopen(input_file, "r");
    if (file == NULL)
    {
        printf("Cannot open file %s\n", input_file);
        success = 0;
    }
    else
    {
        /* read the 2 integers from file into one array element until end
         * of input
         */
        while (1)
        {
            /* read 2 integers */
            int base, digits;
            fscanf(file, "%d %d", &base, &digits);
            if (_debug >= DEBUG_COPIOUS)
               printf("%d %d\n", base, digits);

            /* both integers being zero means the end of input */
            if (base == 0 || digits == 0)
                break;
            else
            {
                lcds[*count]._base = base;
                sprintf(lcds[*count]._digits, "%d", digits);
                (*count)++;
            }
        }
        fclose(file);
        success = 1;
    }

    return success;
}


void output(lcd_t* list, int count)
{
    /* process the actual entries in the array */
    int i;
    for (i = 0; i < count; i++)
    {
        print(list[i]._base, list[i]._digits);
        printf("\n");
    }
}


void main(int argc, char** argv)
{
    char opt;
    char* input_file = NULL;

    while ((opt = getopt(argc, argv, "i:d:")) != -1)
    {
        switch (opt)
        {
        case 'i':
            input_file = optarg;
            break;
        case 'd':
            _debug = atoi(optarg);
            break;
        default:
            printf("Usage: %s -i input_file [-d 0|1|2]\n", argv[0]);
        }
    }

    if (input_file == NULL || _debug < DEBUG_OFF || _debug > DEBUG_COPIOUS)
        printf("Usage: %s -i input_file [-d 0|1|2]\n", argv[0]);
    else
    {
        /* each test case is one line of input, with each line having two
         * numbers representing the base and the digits to be displayed
         */
        lcd_t list[MAX_CASES];
        int count = 0;  /* how many actual entries in the array */
        if (input(input_file, list, &count) != 0)
            output(list, count);
    }
}

File: sol_1_6_4.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.ArrayList;


class sol_1_6_4 {
   private static void input(ArrayList<LCD> list) throws FileNotFoundException {
      Scanner scanner = new Scanner(System.in);
      while (true) {
         // read the 2 integers from file into one array element until end
         // of input
         int base = scanner.nextInt();
         int digits = scanner.nextInt();

         // both integers being zero means the end of input
         if (base == 0 || digits == 0)
            break;
         else
         {
            LCD tmp = new LCD();
            tmp.base(base);
            tmp.digits(digits);

            list.add(tmp);
         }
      }
      scanner.close();
   }


   private static void output(ArrayList<LCD> list) {
      // process the actual entries in the array
      for (int i = 0; i < list.size(); i++)
         System.out.print(list.get(i));
   }


   public static void main(String[] args) throws FileNotFoundException {
      ArrayList<LCD> list = new ArrayList<LCD>();
      input(list);
      output(list);
   }
}


class LCD {
   private int   _base;
   private String _digits;


   public void base(int base) {
      _base = base;
   }


   public void digits(int number) {
      _digits = Integer.toString(number);
   }


   public String toString() {
      Printer printer = new Printer();

      for (int row = 0; row < 2*_base+3; row++) {
         for (int i = 0; i < _digits.length(); i++) {
            String tmp = printer._builder.toString();
            if (i != 0)
               printer._builder.append(" ");
            for (int column = 0; column < _base+2; column++) {
               int digit = _digits.charAt(i) - '0';
               printer._functors[digit].execute(_base, row, column);
            }
         }
         printer._builder.append("\n");
      }
      printer._builder.append("\n");

      return printer._builder.toString();
   }
}


class Printer {
   public Functor[] _functors;
   public StringBuilder _builder = new StringBuilder();


   public Printer() {
      if (_functors == null)
      {
         _functors = new Functor[10];

         _functors[0] = new Functor0();
         _functors[1] = new Functor1();
         _functors[2] = new Functor2();
         _functors[3] = new Functor3();
         _functors[4] = new Functor4();
         _functors[5] = new Functor5();
         _functors[6] = new Functor6();
         _functors[7] = new Functor7();
         _functors[8] = new Functor8();
         _functors[9] = new Functor9();
      }
   }


   public interface Functor {
      public void execute(int base, int row, int column);
   }


   class Functor0 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == 2*base+2)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else if (row == base+1)
            _builder.append(" ");
         else
         {
            if (column == 0 || column == base+1)
               _builder.append("|");
            else
               _builder.append(" ");
         }
      }
   }


   class Functor1 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == base+1 || row == 2*base+2)
            _builder.append(" ");
         else
         {
            if (column == base+1)
               _builder.append("|");
            else
               _builder.append(" ");
         }
      }
   }


   class Functor2 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == base+1 || row == 2*base+2)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else
         {
            if (row <= base)
            {
               if (column == base+1)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
            else
            {
               if (column == 0)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
         }
      }
   }


   class Functor3 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == base+1 || row == 2*base+2)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else
         {
            if (column == base+1)
               _builder.append("|");
            else
               _builder.append(" ");
         }
      }
   }


   class Functor4 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == 2*base+2)
            _builder.append(" ");
         else if (row == base+1)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else
         {
            if (row <= base)
            {
               if (column == 0 || column == base+1)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
            else
            {
               if (column == base+1)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
         }
      }
   }


   class Functor5 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == base+1 || row == 2*base+2)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else
         {
            if (row <= base)
            {
               if (column == 0)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
            else
            {
               if (column == base+1)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
         }
      }
   }


   class Functor6 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == base+1 || row == 2*base+2)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else
         {
            if (row <= base)
            {
               if (column == 0)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
            else
            {
               if (column == 0 || column == base+1)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
         }
      }
   }


   class Functor7 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else if (row == base+1 || row == 2*base+2)
            _builder.append(" ");
         else
         {
            if (column == base+1)
               _builder.append("|");
            else
               _builder.append(" ");
         }
      }
   }


   class Functor8 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == base+1 || row == 2*base+2)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else
         {
            if (column == 0 || column == base+1)
               _builder.append("|");
            else
               _builder.append(" ");
         }
      }
   }


   class Functor9 implements Functor {
      public void execute(int base, int row, int column) {
         String tmp = _builder.toString();
         if (row == 0 || row == base+1 || row == 2*base+2)
         {
            if (column == 0 || column == base+1)
               _builder.append(" ");
            else
               _builder.append("-");
         }
         else
         {
            if (row <= base)
            {
               if (column == 0 || column == base+1)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
            else
            {
               if (column == base+1)
                  _builder.append("|");
               else
                  _builder.append(" ");
            }
         }
      }
   }
}
